videoretrieval/cache/language_model_cache.py

```diff
@@ -0,0 +1,159 @@
+            model used to generate the contextual embeddings.
+        split: the name of the split (as a string).
+
+    Returns: a tf.data Dataset where the first element is the video id as a
+        string tensor, the second is the contextual embeddings as a float32
+        tensor, and the third is the raw caption text as a string tensor.
+        The dataset is empty if there are no cached results.  
+    """
+
+    dataset = dataset.unbatch().map(seralize_to_protobuf_wrapper, 
+        num_parallel_calls=tf.data.experimental.AUTOTUNE)
+
+    records_directory = get_records_directory(source_dataset, language_model,
+        split)
+
+    write_dataset(dataset, records_directory, 
+        source_dataset.num_of_examples_by_split(split))
+
+def get_cached_records(source_dataset, language_model, split):
+    """Get a list of files that contain cached data.
+
+    Parameters:
+        source_dataset: an implementation of BaseDataset that the cached
+            embeddings were generated from. 
+        language_model: an implementation of BaseLanguage model that the
+            cached embeddings were generated from.
+        split: the name of the dataset split.
+
+    Returns: a list of file paths (as strings) to each cached tfrecord file.
+    """
+
+    records_directory = get_records_directory(
+        source_dataset, language_model, split)
+    glob_string = str((records_directory / "*.tfrecord").absolute())
+
+    return glob.glob(glob_string)
+
+def unseralize_data(seralized_item):
+    """Unseralizes a seralized protobuf feature.
+
+    Returns: a tuple of 3 items, the first being the video id as a string tensor,
+        the second being the contextual embeddings as a float32 tensor,
+        the third being the source text as tensor.
+    """
+    example = tf.io.parse_single_example(seralized_item, decoding_schema)
+    contextual_embeddings = tf.io.parse_tensor(
+        example["seralized_embeddings"], tf.float32)
+
+    return (example["video_id"], contextual_embeddings, example["text"])
+
+def get_cached_language_model_embeddings(source_dataset, language_model, split):
+    """Load the cached embeddings for a specific dataset/model/split.
+
+    Arguments:
+        source_dataset: an implementation of BaseDataset for the dataset to be
+            loaded.
+        language_model: an implementation of BaseLanguageModel for the language
+            model used to generate the contextual embeddings.
+        split: the name of the split (as a string).
+
+    Returns: a tf.data Dataset where the first element is the video id as a
+        string tensor, the second is the contextual embeddings as a float32
+        tensor, and the third is the raw caption text as a string tensor.
+        The dataset is empty if there are no cached results.  
+    """
+    record_files = get_cached_records(source_dataset, language_model, split)
+
+    return (tf.data.TFRecordDataset(record_files)
```

Comment:
I've made a bunch of performance improvements in a different branch, so the code on this branch is not the most optimzied.